Test services examples

esteve opened the pull request:
Adds tests for services

Depends on #74 

Connects to ros2/ros2#143

esteve commented:
CI jobs:

http://ci.ros2.org/job/ros2_batch_ci_linux/618/
http://ci.ros2.org/job/ros2_batch_ci_osx/509/
http://ci.ros2.org/job/ros2_batch_ci_windows_opensplice/685/

dirk-thomas commented on rclcpp_examples/CMakeLists.txt:
```diff
@@ -101,23 +101,36 @@ if(AMENT_ENABLE_TESTING)
-  foreach(example_executable list_parameters_async list_parameters parameter_events_async parameter_events set_and_get_parameters_async set_and_get_parameters)
+  # Add each test case.  Multi-executable tests can be specified in
+  # semicolon-separated strings, like  exe1;exe2.
+  foreach(example_executable list_parameters_async list_parameters parameter_events_async parameter_events set_and_get_parameters_async set_and_get_parameters "talker;listener" "add_two_ints_client;add_two_ints_server" "add_two_ints_client_async;add_two_ints_server")
```
This line seems to become excessively long. Can you create a list variable for this before and place every entry on a separate line to make it more readable?

esteve replied:
Done.

dirk-thomas commented on rclcpp_examples/test/test_executables_example.py.in:
```diff
@@ -0,0 +1,59 @@
+            name=name,
+            exit_handler=ignore_exit_handler,
+            output_handlers=[handler],
+        )
```
Can some of this be provided by `launch_testing`? Otherwise it will likely get duplicated in other tests.

esteve replied:
I've moved some of this code to https://github.com/ros2/launch/pull/15

esteve commented:
I've addressed the feedback for this PR, and the new tests seem to pass on the Linux and OSX CI jobs:

http://ci.ros2.org/job/ci_linux/626/
http://ci.ros2.org/job/ci_osx/515/

The fix for Windows is in https://github.com/osrf/osrf_pycommon/pull/17

dirk-thomas commented on rclcpp_examples/CMakeLists.txt:
```diff
@@ -103,8 +103,18 @@ if(AMENT_ENABLE_TESTING)
+    set_and_get_parameters
+    "talker\;listener"
+    "add_two_ints_client\;add_two_ints_server"
+    "add_two_ints_client_async\;add_two_ints_server")
```
While using `\;` here works it is very fragile. E.g. assigning the variable `example_executables` to `example_executables2` would already break it. Therefore I would suggest to use a different separate (e.g. `:`) and replace it within the first loop with a `semicolon`.

esteve replied:
I've replaced the code with this, but I can't get it to work:

```
  # Add each test case.  Multi-executable tests can be specified in
  # semicolon-separated strings, like  exe1;exe2.
  set(example_executables
    list_parameters_async
    list_parameters
    parameter_events_async
    parameter_events
    set_and_get_parameters_async
    set_and_get_parameters
    "talker:listener"
    "add_two_ints_client:add_two_ints_server"
    "add_two_ints_client_async:add_two_ints_server")

  foreach(example_executable_raw ${example_executables})
    set(REPLACE ";" ":" "${example_executable_raw}" example_executable)
    set(RCLCPP_EXAMPLES_EXPECTED_OUTPUT "")
    foreach(executable ${example_executable})
```

what'd be the correct approach?

esteve replied:
I just realized the problem is in the second `set` command

esteve replied:
Changed this to use ":" and then replace it with ";"

esteve commented:
I've squashed the commits, this should be ready for merging.

esteve commented:
I started a new CI job for Windows:

http://ci.ros2.org/job/ci_windows_opensplice/699/

dirk-thomas commented:
The variable names are kind of confusing. The outer list sounds more like `tests` while the inner one is `executables` (still plural).

Beside that: lgtm.

esteve commented:
@dirk-thomas I've renamed the variables for the tests and the executables.

dirk-thomas commented on rclcpp_examples/CMakeLists.txt:
```diff
@@ -103,8 +103,19 @@ if(AMENT_ENABLE_TESTING)
+    "add_two_ints_client_async:add_two_ints_server")
+
+  foreach(example_test ${example_tests})
+    string(REPLACE ":" ";" example_executable "${example_test}")
```
Since `example_executable` is a list it should be plural.

esteve replied:
Fixed.
